### PR TITLE
KAFKA-8753: expose the number of topics marked for deletion

### PR DESCRIPTION
**Pull request by mumrah**
This is the implementation for [KIP-503](https://cwiki.apache.org/confluence/display/KAFKA/KIP-503%3A+Add+metric+for+number+of+topics+marked+for+deletion)

When deleting a large number of topics, the Controller can get quite bogged down. One problem with this is the lack of visibility into the progress of the Controller. We can look into the ZK path for topics marked for deletion, but in a production environment this is inconvenient. This PR adds a JMX metric `kafka.controller:type=KafkaController,name=TopicsToDeleteCount` to make it easier to see how many topics are being deleted.

### CODE REVIEW

**Comment by ijuma** [comment]
Thanks for the PR. This requires a KIP since metrics are public API.

**Comment by mumrah** [comment]
Created [KIP-503](https://cwiki.apache.org/confluence/display/KAFKA/KIP-503%3A+Add+metric+for+number+of+topics+marked+for+deletion)

**Comment by mumrah** [comment]
retest this please

File: core/src/test/scala/unit/kafka/metrics/MetricsTest.scala [hunk 1]
```diff
@@ -141,6 +141,7 @@ class MetricsTest extends KafkaServerTestHarness with Logging {
     assertEquals(metrics.keySet.asScala.count(_.getMBeanName == "kafka.controller:type=KafkaController,name=PreferredReplicaImbalanceCount"), 1)
     assertEquals(metrics.keySet.asScala.count(_.getMBeanName == "kafka.controller:type=KafkaController,name=GlobalTopicCount"), 1)
     assertEquals(metrics.keySet.asScala.count(_.getMBeanName == "kafka.controller:type=KafkaController,name=GlobalPartitionCount"), 1)
+    assertEquals(metrics.keySet.asScala.count(_.getMBeanName == "kafka.controller:type=KafkaController,name=TopicsToDeleteCount"), 1)
```
Comment (stanislavkozlovski) [inline core/src/test/scala/unit/kafka/metrics/MetricsTest.scala:144]: Should we also add the `replicasToDeleteCount` here?

**Comment by stanislavkozlovski** [comment]
Happy to see the replica count metric! Let's update the KIP to mention it as well

**Review by junrao — APPROVED** [review]
@mumrah : Thanks for the PR. LGTM assuming that the KIP will pass.

File: core/src/main/scala/kafka/controller/KafkaController.scala [hunk 2]
```diff
@@ -1191,6 +1227,27 @@ class KafkaController(val config: KafkaConfig,
...
+
+    topicsToDeleteCount = if (!isActive) 0 else {
+      (controllerContext.topicsToBeDeleted -- controllerContext.topicsIneligibleForDeletion).size
+    }
+
+    replicasToDeleteCount = if (!isActive) 0 else controllerContext.topicsToBeDeleted.map { topic =>
+      // For each enqueued topic, count the number of eligible replicas that are not yet deleted
+      controllerContext.replicasForTopic(topic).count { replica =>
+        controllerContext.replicaState(replica) != ReplicaDeletionSuccessful &&
+        controllerContext.replicaState(replica) != ReplicaDeletionIneligible
```
Comment (hachikuji) [inline core/src/main/scala/kafka/controller/KafkaController.scala]: Hmm.. Why do we exclude replicas that are currently ineligible? I think eventually they still need to get deleted.

**Review by hachikuji — COMMENTED** [review]
Thanks, left one comment.

File: core/src/main/scala/kafka/controller/KafkaController.scala (hunk 2 above)
Comment (mumrah) [reply core/src/main/scala/kafka/controller/KafkaController.scala]: The idea is that the two counts (replicasToDelete and ineligibleReplicasToDelete) are mutually exclusive so they can be combined to see the total number of pending replica deletes. This was a suggestion in the KIP discussion. Alternatively, we could have replicasToDeleteCount represent _all_ pending replica deletes regardless of eligibility.

Comment (hachikuji) [reply core/src/main/scala/kafka/controller/KafkaController.scala]: Hmm, I see. I guess that's reasonable. It seems to make the naming a little less clear though. For example, `TopicsToDeleteCount` really means something like `EligibleTopicsToDeleteCount`. I think I slightly favor letting `TopicsToDeleteCount` just represent the count of topics that are being deleted. Similarly for replicas.

Comment (mumrah) [reply core/src/main/scala/kafka/controller/KafkaController.scala]: Yea, i actually kind of like prefer that as well. Updated.

**Comment by hachikuji** [comment]
@mumrah Thanks, the patch looks good. Do you think it's worthwhile mentioning these metrics in `docs/ops.html`?

**Review by hachikuji — APPROVED** [review]
LGTM

**Pull request merged**